src/data/repositories/UserD2ApiRepository.ts

```diff
@@ -469,52 +470,64 @@ export class UserD2ApiRepository implements UserRepository {
         const allExistingUsersGroups = this.buildUsersByGroupId(existing);
 
-        const existingKeys = _(allExistingUsersGroups).keys().value();
+        const userGroupsWithUsersToAdd = getDiffUserIdsByGroup(allUsersGroupsToUpdate, allExistingUsersGroups);
+        const userGroupsWithUsersToRemove = getDiffUserIdsByGroup(allExistingUsersGroups, allUsersGroupsToUpdate);
 
-        const groupsIdsToAddRef = users.flatMap(user => {
-            const groupsRef = user.userGroups.map(userGroup => ({ id: userGroup.id }));
-            return groupsRef.filter(({ id }) => !existingKeys.includes(id));
-        });
+        const $requestsToAdd = this.buildRequestsGroups(userGroupsWithUsersToAdd, "add");
+        const $requestsToDelete = this.buildRequestsGroups(userGroupsWithUsersToRemove, "delete");
 
-        const groupsIdsToAdd = _.uniqBy(groupsIdsToAddRef, ({ id }) => id);
+        return Future.sequential([$requestsToAdd, $requestsToDelete]).flatMap(() => {
+            const groupsIdsToAdd = userGroupsWithUsersToAdd
+                .filter(group => group.usersIds.length > 0)
+                .map(group => group.id);
+            const groupsIdsToDelete = userGroupsWithUsersToRemove
+                .filter(group => group.usersIds.length > 0)
+                .map(group => group.id);
 
-        const groupsIdsToDelete = users.flatMap(user => {
-            const existingUser = existing.find(({ id }) => id === user.id);
-            const difference = _.differenceWith(
-                existingUser?.userGroups,
-                user.userGroups,
-                (user1, user2) => user1.id === user2.id
-            );
-            return difference.map(userGroup => ({ id: userGroup.id }));
-        });
+            if (logger) {
+                this.logGroupChanges(logger, userGroupsWithUsersToAdd, "add");
+                this.logGroupChanges(logger, userGroupsWithUsersToRemove, "delete");
 
-        const $requestsToAdd = this.buildRequestsGroups(groupsIdsToAdd, allUsersGroups, "add");
-        const $requestsToDelete = this.buildRequestsGroups(groupsIdsToDelete, allExistingUsersGroups, "delete");
+                logger.log({ groupsIdsToAdd: groupsIdsToAdd, groupsIdsToDelete: groupsIdsToDelete });
+            }
+
+            return Future.success(undefined);
+        });
+    }
 
-        return Future.sequential([...$requestsToAdd, ...$requestsToDelete]).map(stats => {
-            logger?.log({ groupsIdsToAdd: groupsIdsToAdd, groupsIdsToDelete: groupsIdsToDelete });
-            return Stats.combine(stats);
+    private logGroupChanges(
+        logger: D2LoggerMessage,
+        groups: Array<{
+            id: Id;
+            usersIds: Id[];
+        }>,
+        action: "add" | "delete"
+    ) {
+        groups.forEach(group => {
+            if (group.usersIds.length > 0) {
+                logger.log({
+                    action: action,
+                    groupId: group.id,
+                    usersIds: group.usersIds,
```

Comment:
The orthodox here is "userIds", two plurals sound weird. Change only if it we don't have already data saved, then it's a good time to do it.